ui/gtk4.c

```diff
@@ -0,0 +1,2176 @@
+        qemu_input_queue_rel(vc->gfx.dcl.con, INPUT_AXIS_X, x - s->last_x);
+        qemu_input_queue_rel(vc->gfx.dcl.con, INPUT_AXIS_Y, y - s->last_y);
+        qemu_input_event_sync();
+    }
+    s->last_x = x;
+    s->last_y = y;
+    s->last_set = TRUE;
+
+    if (!qemu_input_is_absolute() && s->ptr_owner == vc) {
+        GdkDisplay *dpy = gtk_widget_get_display(vc->gfx.drawing_area);
+        GtkNative *native = gtk_widget_get_native(vc->gfx.drawing_area);
+        GdkSurface *surface = gtk_native_get_surface(native);
+        GdkMonitor *monitor = gdk_display_get_monitor_at_surface(dpy, surface);
+        GdkRectangle geometry;
+
+        int x = (int)pointer_x;
+        int y = (int)pointer_y;
+
+        gdk_monitor_get_geometry(monitor, &geometry);
+
+        /*
+         * In relative mode check to see if client pointer hit
+         * one of the monitor edges, and if so move it back to the
+         * center of the monitor. This is important because the pointer
+         * in the server doesn't correspond 1-for-1, and so
+         * may still be only half way across the screen. Without
+         * this warp, the server pointer would thus appear to hit
+         * an invisible wall
+         */
+        if (x <= geometry.x || x - geometry.x >= geometry.width - 1 ||
+            y <= geometry.y || y - geometry.y >= geometry.height - 1) {
+            x = geometry.x + geometry.width / 2;
+            y = geometry.y + geometry.height / 2;
+
+            /*
+             * TODO: replace with X11 spexific API
+             * GdkScreen *screen = gtk_widget_get_screen(vc->gfx.drawing_area);
+             * GdkDevice *dev = gdk_event_get_device((GdkEvent *)motion);
+             * gdk_device_warp(dev, screen, x, y);
+             */
+            s->last_set = FALSE;
+            return ;
+        }
+    }
+    return ;
+}
+
+static void gd_button_event(GtkEventController *controller, gint n_presses,
+                            double x, double y, VirtualConsole *vc)
+{
+    GdkEvent *button = gtk_event_controller_get_current_event(controller);
+    GtkDisplayState *s = vc->s;
+    guint button_nbr = gdk_button_event_get_button(GDK_EVENT(button));
+    InputButton btn;
+
+    /* implicitly grab the input at the first click in the relative mode */
+    if (button_nbr == 1 && n_presses == 1 &&
+        !qemu_input_is_absolute() && s->ptr_owner != vc) {
+        if (!vc->window) {
+            s_action_set_state(s->grab_action, TRUE);
+        } else {
+            gd_grab_pointer(vc, "relative-mode-click");
+        }
+    }
+
+    if (button_nbr == 1) {
+        btn = INPUT_BUTTON_LEFT;
+    } else if (button_nbr == 2) {
+        btn = INPUT_BUTTON_MIDDLE;
+    } else if (button_nbr == 3) {
+        btn = INPUT_BUTTON_RIGHT;
+    } else if (button_nbr == 8) {
+        btn = INPUT_BUTTON_SIDE;
+    } else if (button_nbr == 9) {
+        btn = INPUT_BUTTON_EXTRA;
+    } else {
+        return;
+    }
```

Comment:
This feels like a switch statement would be a better fit. And instead of all the magic numbers, maybe use an `enum` to check the value?

Reply:
I suppose those are magic numbers for a reason, even in upstream GTK, otherwise they would have been an enum there as well. I switched to a switch nonetheless